core/src/util/dsn.rs

```diff
@@ -0,0 +1,308 @@
+        MAX_VALUE_LENGTH
+    )]
+    Value,
+    #[error("{}", .0)]
+    Generic(String),
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct DSNOpts {
+    pub database: String,
+    pub dsn: String,
+    pub password: String,
+    pub server: String,
+    pub user: String,
+    // SQL-1281
+    // pub logpath: String,
+    pub driver_name: String,
+}
+
+impl DSNOpts {
+    pub fn new(
+        database: String,
+        dsn: String,
+        password: String,
+        server: String,
+        user: String,
+        driver_name: String,
+    ) -> Result<Self, DSNError> {
+        match (
+            DSNOpts::check_value_length(&database),
+            unsafe { SQLValidDSNW(to_widechar_ptr(&dsn).0) },
+            DSNOpts::check_value_length(&password),
+            DSNOpts::check_value_length(&server),
+            DSNOpts::check_value_length(&user),
+            DSNOpts::check_value_length(&driver_name),
+        ) {
+            (true, true, true, true, true, true) => Ok(Self {
+                database,
+                dsn,
+                password,
+                server,
+                user,
+                driver_name,
+            }),
+            (_, false, _, _, _, _) => Err(DSNError::DSN(dsn)),
+            _ => Err(DSNError::Value),
+        }
+    }
+    pub fn from_attribute_string(attribs: String) -> Option<Self> {
+        match ODBCUri::new(&attribs.replace(char::from(0), ";")) {
+            Ok(uri) => Some(Self::from(uri)),
+            Err(_) => None,
+        }
+    }
+
+    pub fn is_valid_dsn(&self) -> bool {
+        unsafe { SQLValidDSNW(to_widechar_ptr(&self.dsn).0) }
+    }
+
+    pub fn write_dsn_to_registry(&self) -> bool {
+        (unsafe {
+            SQLWriteDSNToIniW(
+                to_widechar_ptr(&self.dsn).0,
+                to_widechar_ptr(&self.driver_name).0,
+            )
+        }) && self.write_private_profile_string()
+    }
+
+    fn write_private_profile_string(&self) -> bool {
+        self.iter().all(|(key, value)| unsafe {
+            SQLWritePrivateProfileStringW(
+                to_widechar_ptr(&self.dsn).0,
+                to_widechar_ptr(key).0,
+                to_widechar_ptr(value).0,
+                to_widechar_ptr(ODBCINI).0,
+            )
+        })
+    }
+
+    pub fn from_private_profile_string(&self) -> Self {
+        let buffer = &mut [0u16; MAX_VALUE_LENGTH];
+        let mut dsn_opts = DSNOpts::default();
+        self.iter().for_each(|(key, _)| {
+            let len = unsafe {
+                SQLGetPrivateProfileStringW(
+                    to_widechar_ptr(&self.dsn).0,
+                    to_widechar_ptr(key).0,
+                    to_widechar_ptr("").0,
+                    buffer.as_mut_ptr(),
+                    buffer.len() as i32,
+                    to_widechar_ptr(ODBCINI).0,
+                )
+            };
+            let value = unsafe { input_text_to_string_w(buffer.as_mut_ptr(), len as usize) };
```

Comment:
I will add a hard check. I can't safeguard against it, but it should be impossible based on Microsoft's registry value documentation. They could change it in the future, which could cause this to fail. 🤔 

This may be a good place for a debug write, with something like "If you see this error please report it."

Reply:
Yes, I think a debug write with the key name could be useful. But unless they have a bunch of giant values, it should easy to figure out which key is problematic